src/Microsoft.ML.Data/Data/Conversion.cs

```diff
@@ -1333,11 +1305,11 @@ public bool TryParse(ref TX src, out I2 dst)
         /// </summary>
         public bool TryParse(ref TX src, out I4 dst)
         {
-            long res;
-            bool f = TryParseSigned(RawI4.MaxValue, ref src, out res);
-            Contracts.Assert(f || res == I4.RawNA);
-            Contracts.Assert((RawI4)res == res);
-            dst = (RawI4)res;
+            long? res;
+            bool f = TryParseSigned(I4.MaxValue, ref src, out res);
+            Contracts.Check(f && res.HasValue);
```

Comment:
>Contracts.Check(f && res.HasValue); [](start = 12, length = 35)

This needs some work. If you're changing these to be user facing error messages you're going to need to provide a sensible error message.

You are also providing a failure to parse in two distinct ways now, both by returning false from the `TryParse` method, as well as returning a null value in `res`. That seems needlessly inefficient.